rs/node_rewards/canister/src/test.rs

```diff
@@ -0,0 +1,236 @@
+use crate::canister::NodeRewardsCanister;
+use crate::storage::RegistryStoreStableMemoryBorrower;
+use futures_util::FutureExt;
+use ic_nervous_system_canisters::registry::{
+    FakeRegistry, FakeRegistryResponses, RegistryCanister,
+};
+use ic_node_rewards_canister_api::monthly_rewards::{
+    GetNodeProvidersMonthlyXdrRewardsRequest, GetNodeProvidersMonthlyXdrRewardsResponse,
+    NodeProvidersMonthlyXdrRewards,
+};
+use ic_protobuf::registry::dc::v1::DataCenterRecord;
+use ic_protobuf::registry::node_operator::v1::NodeOperatorRecord;
+use ic_protobuf::registry::node_rewards::v2::{NodeRewardRate, NodeRewardRates, NodeRewardsTable};
+use ic_registry_canister_client::{registry_data_stable_memory_impl, StableCanisterRegistryClient};
+use ic_registry_canister_client::{
+    CanisterRegistryClient, RegistryDataStableMemory, StorableRegistryKey, StorableRegistryValue,
+};
+use ic_registry_keys::{
+    make_api_boundary_node_record_key, make_data_center_record_key, make_node_operator_record_key,
+    NODE_REWARDS_TABLE_KEY,
+};
+use ic_registry_node_provider_rewards::logs::RewardsPerNodeProviderLog;
+use ic_registry_node_provider_rewards::RewardsPerNodeProvider;
+use ic_registry_transport::pb::v1::{RegistryDelta, RegistryValue};
+use ic_stable_structures::memory_manager::{MemoryId, MemoryManager, VirtualMemory};
+use ic_stable_structures::{DefaultMemoryImpl, StableBTreeMap};
+use ic_types::{PrincipalId, RegistryVersion};
+use maplit::btreemap;
+use std::cell::RefCell;
+use std::str::FromStr;
+use std::sync::Arc;
+
+pub type VM = VirtualMemory<DefaultMemoryImpl>;
+
+thread_local! {
+    static STATE: RefCell<StableBTreeMap<StorableRegistryKey, StorableRegistryValue, VM>> = RefCell::new({
+        let mgr = MemoryManager::init(DefaultMemoryImpl::default());
+        StableBTreeMap::init(mgr.get(MemoryId::new(0)))
+    });
+}
+
+registry_data_stable_memory_impl!(TestState, STATE);
+
+fn setup_canister_for_test(
```

Comment:
setup function isn't very good at the moment... some details are injected but have to be (confusingly) kept in sync across 2 functions and there is still some assumptions about the state of the Registry locally